# Models of sequence evolution

Before starting you should have an aligned & filtered MSA! you can find some [here](https://github.com/for-giobbe/MP25/tree/main/data/example_alignements)

For all ML analyses we will use IQ-TREE. Some of the greates advantages of this software are:

 * It can do almost everything (take a look at the [manual](http://www.iqtree.org/doc/)).
 * Each analysis can be highly customizable.
 * Really faster compared to other softwares.
 * Very easy to run as you want (if you study a bit the manual).

For a *quick-and-dirty* Model Selection on an MSA we can just type:

```
iqtree2 -s data/example_alignements/N0.HOG0000096.ref.aln -m TESTONLY
```
You can see that it instantly prints some usefull stats which reconnect to concepts we stubled upon earlier - parsimony-informative sites! Some other are new, like distinct patterns, which refers to the unique configurations of nucleotides or amino acids observed across all sequences at specific alignment positions. The test performed here indicates no significant compositional heterogeneity among the sequences, this is somethin which we will get back to in the lesson on biases.

```
Alignment most likely contains protein sequences
WARNING: 2 sites contain only gaps or ambiguous characters.
Alignment has 8 sequences with 503 columns, 358 distinct patterns
109 parsimony-informative, 220 singleton sites, 174 constant sites
                Gap/Ambiguity  Composition  p-value
   1  A_granulata      23.06%    passed    100.00%
   2  B_glabrata       24.85%    passed     99.23%
   3  C_virginica      23.06%    passed     99.90%
   4  H_robusta         9.34%    passed     66.99%
   5  L_gigantea       22.86%    passed    100.00%
   6  O_bimaculoides   24.85%    passed     98.19%
   7  P_fucata         30.82%    passed     99.82%
   8  S_constricta     24.06%    passed     99.88%
****  TOTAL            22.86%  0 sequences failed composition chi2 test (p-value<5%; df=19)
```

You can appreciate from what is shown in the standard outpout that the ptogram is *doing a tree before model selection*! And it does use both parsimony and likelihood ... However, when the models started to be tested, and by now you should have a grasp of what the text here means!

```

 No. Model         -LnL         df  AIC          AICc         BIC
  1  LG            4783.673     13  9593.346     9594.090     9648.214
  2  LG+I          4762.582     14  9553.164     9554.025     9612.252
  3  LG+G4         4730.778     14  9489.556     9490.416     9548.644
  4  LG+I+G4       4731.634     15  9493.267     9494.253     9556.576
  7  LG+F+G4       4680.996     33  9427.992     9432.777     9567.272
  8  LG+F+I+G4     4681.699     34  9431.398     9436.484     9574.898
```

The ```-m TESTONLY``` word stands for  standard model selection, which tells IQ-TREE to perform ModelFinder without taking into consideration FreeRate model (look at them [here](https://www.nature.com/articles/nmeth.4285). You should just add ```-m MF ``` flag to compute a full Model Selection): this tool computes the log-likelihoods of an initial parsimony tree for many different models and the Akaike information criterion (AIC), corrected Akaike information criterion (AICc), and the Bayesian information criterion (BIC). Then ModelFinder chooses the model that minimizes the BIC score (you can also change to AIC or AICc by adding the option ```-AIC``` or ```-AICc```, respectively)

The output are:

 * **My_Concat.fa.ckp.gz:** checkpoint file.
 * **My_Concat.fa.iqtree:** summary human readble file.
 * **My_Concat.fa.log:** log file.
 * **My_Concat.fa.model.gz** computer readble result file.
 * **My_Concat.fa.treefile** Tree used for ModelFinder.

Let's have a look at the iqtree summary file, where we can find a lot of usefull informations:

```
cat data/example_alignements/N0.HOG0000096.ref.aln.iqtree
```


- Q.insect: This denotes a user-defined or empirical amino acid substitution matrix obtained  from insect sequences. It is expected to better represent the evolutionary patterns of particular groups, but  they are just numbers. Do not be surprise when Q.plant is your best model for an animal!

- +F: This component indicates that the model incorporates empirical amino acid frequencies, calculated directly from the provided alignment data. Alternatives are: +FQ	= equal base frequencies; +FO	= optimized base frequencies by maximum-likelihood.

- +G4: This signifies the use of a discrete Gamma distribution with four rate categories to model rate heterogeneity across sites.Additionally: +I	allows for a proportion of invariable sites and +R is the FreeRate model that generalizes the +G model by relaxing the assumption of Gamma-distributed rates.


>The -m flag can also specify a model name to use during the analyses, which can be a priori specified by the user (here's a [list](http://www.iqtree.org/doc/Substitution-Models) of models implemented in ModelFinder). Usually this is done to save computational time when the best-fit model has been already pre-computed.


---


### Quiz! 
- Is JTT or LG a better model for our alignement?
- [Here](https://github.com/for-giobbe/MP25/blob/main/data/example_alignements/Rhabdomeric1.nt.aln) you can find a nucleotide alignement ... can you tell me the probability of A↔C?! Gan you tell me the alpha parameter of the gamma distribution?


__For experienced phylogeneticists only:__

>​In phylogenetic substitution models, the Rate Parameters (R) and the Rate Matrix (Q) are closely related but different. The >Rate Parameters (R) represent the relative substitution rates between different nucleotide pairs, while the Rate Matrix (Q) >incorporates these rates along with nucleotide frequencies to describe the overall substitution process. The differences in >values between R and Q arise due to the incorporation of nucleotide frequencies into the Q matrix.​
>
>Rate Matrix (Q): This matrix incorporates the rate parameters (R) and the equilibrium nucleotide frequencies (π) to describe >the substitution process. Each off-diagonal element qᵢⱼ of the Q matrix is calculated as:​
>
>qᵢⱼ = rᵢⱼ × πⱼ
>
>where rᵢⱼ is the relative rate parameter for the substitution from nucleotide i to nucleotide j, and πⱼ is the equilibrium >frequency of nucleotide j.
>
>The diagonal elements qᵢᵢ are set such that each row sums to zero, ensuring the proper probabilistic behavior of the model.
>
>Example Calculation:
>
>Assuming equal nucleotide frequencies (πₐ = π꜀ = π₉ = πₜ = 0.25), the off-diagonal elements of the Q matrix can be calculated >as:​
>
>A→C: qₐ꜀ = rₐ꜀ × π꜀ = 1.6385 × 0.25 = 0.4096​
>A→G: qₐ₉ = rₐ₉ × π₉ = 3.0090 × 0.25 = 0.7523​
>A→T: qₐₜ = rₐₜ × πₜ = 1.6385 × 0.25 = 0.4096​
>
>The diagonal element qₐₐ is then calculated to ensure the row sums to zero:​
>
>qₐₐ = -(qₐ꜀ + qₐ₉ + qₐₜ) = -(0.4096 + 0.7523 + 0.4096) = -1.5715​
>
>This process is repeated for all nucleotides to construct the full Q matrix.
>
>