
After having aligned and filtere our genes, we will concatenate them using AMAS, a nice Python toolkit. However, as preliminar steps we need to:
  * Convert the gblock format into standard fasta. 
  * Change all fasta headers leaving only the species name. This step is mandatory beacuse concatenation is **always** made looking at the headers.
  * Create a new working directory

```
mkdir Analyses/IQ-TREE
for i in Analyses/1-to-1_Alignments/gblock/*gb; do
	varOG=$( echo "$i" | cut -d"." -f1 | cut -d"/" -f4); cat "$i" | tr -d " " | sed 's/_1//g' > Analyses/1-to-1_Alignments/gblock/"$varOG".t_cofee-gb.fa;
done

for i in Analyses/1-to-1_Alignments/gblock/*.fa; do
	sed -i.old 's/|.*$//g' "$i";
	rm Analyses/1-to-1_Alignments/gblock/*.old;
done
```



for i in *fa; do mafft $i > ${i:0:${#i}-2}aln; done

for i in *aln; do awk -F "|" '{print $1}' $i >  ${i:0:${#i}-3}ref.aln; done