#  mixture (and complex) models 

Traditional model relyes on the assumption that evolution of DNA sequences follow a markov process. However, again, it may not **always** be true. For this reason IQ-TREE implements three *matched-pairs tests of symmetry* ([Naser-Khdour et al., 2019](https://academic.oup.com/gbe/article/11/12/3341/5571717)) to test the two assumptions of *stationarity* and *homogeneity*. Morever are also implemented some Lie Markov models ([Woodhams et al., 2015](https://academic.oup.com/sysbio/article/64/4/638/1650486)), many of which are non-reversible models.

Finally, in Bayesian Inference in the latest years has become very fashionable the CAT model. This mixture model allows **each position** on the alignment to evolve differently from the others, whithout needing of any *a priori* knowledge (*i.e* partitioning scheme). IQ-TREE implements a ML variant called **GHOST**([Crotty et al., 2020](https://academic.oup.com/sysbio/article-abstract/69/2/249/5541793?redirectedFrom=fulltext)). However they are very parameters rich and can only be used when analyzing a lot of data (*i.e* phylogenomic).