
The previous analysis will result in separate models for each partion. Nonetheless, there are several reasons for which we wanto to **merge partitions** which can be described by similar models of evolution, possibly including a better estimation of model parameters (less parameters to estimates = better estimation). ModelFinder implements a *greedy* strategy ([Lanfear et al., 2012](https://academic.oup.com/mbe/article/29/6/1695/1000514)) that starts with the full partition model and subsequentially merges two genes until the model fit does not increase any further. Since this process can be very long when analyszing a lot of partitions, ModelFinder implements the *relaxed hierarchical clustering algorithm* ([Lanfear et al., 2014](https://bmcecolevol.biomedcentral.com/articles/10.1186/1471-2148-14-82)), which is invoked via ```-rcluster``` option and can save a great amount of computational time.

To carry out simultaneously model of evolution & partitioning scheme selection let's use:

```
iqtree -s Analyses/IQ-TREE/My_Concat.fa -spp Analyses/IQ-TREE/partitions.txt -m TESTMERGEONLY --prefix Analyses/IQ-TREE/spp_MergedPartitions -rcluster 10
```

where with ```-rcluster 10``` we are only examining the top 10% partition merging schemes (better to carry on a full *greedy* search in real scenarios).

Now we can have a look at the resulting partitioning scheme and its BIC score:

```
cat Analyses/IQ-TREE/spp_MergedPartitions.best_scheme.nex
#nexus
begin sets;
  charset p1_OG0000030_p11_OG0000040_p41_OG0000070 = 1-197  3704-4054  20714-21356;
  charset p2_OG0000031_p3_OG0000032_p7_OG0000036_p9_OG0000038_p13_OG0000042_p16_OG0000045_p22_OG0000051_p26_OG0000055_p32_OG0000061_p33_OG0000062_p48_OG0000077_p49_OG0000078 = 198-348  349-1070  2301-2463  2868-3281  5125-5285  6215-7227  9914-10832  11579-11838  17913-18242  18243-18436  24009-24208  24209-24546;
  charset p4_OG0000033_p10_OG0000039_p12_OG0000041_p34_OG0000063 = 1071-1512  3282-3703  4055-5124  18437-18757;
  charset p5_OG0000034_p15_OG0000044_p17_OG0000046_p18_OG0000047_p21_OG0000050_p23_OG0000052_p25_OG0000054_p27_OG0000056_p35_OG0000064_p36_OG0000065_p38_OG0000067_p39_OG0000068_p43_OG0000072 = 1513-1925  5483-6214  7228-7383  7384-7908  9495-9913  10833-10985  11444-11578  11839-12219  18758-18953  18954-19237  19608-19890  19891-20275  21798-22249;
  charset p6_OG0000035_p30_OG0000059 = 1926-2300  17123-17421;
  charset p8_OG0000037_p19_OG0000048_p42_OG0000071_p47_OG0000076 = 2464-2867  7909-9018  21357-21797  23553-24008;
  charset p14_OG0000043_p31_OG0000060_p46_OG0000075 = 5286-5482  17422-17912  23402-23552;
  charset p20_OG0000049_p45_OG0000074 = 9019-9494  23002-23401;
  charset p24_OG0000053 = 10986-11443;
  charset p28_OG0000057_p29_OG0000058 = 12220-16792  16793-17122;
  charset p37_OG0000066_p40_OG0000069_p44_OG0000073 = 19238-19607  20276-20713  22250-23001;
  charpartition mymodels =
    LG+I+G4: p1_OG0000030_p11_OG0000040_p41_OG0000070,
    LG+F+G4: p2_OG0000031_p3_OG0000032_p7_OG0000036_p9_OG0000038_p13_OG0000042_p16_OG0000045_p22_OG0000051_p26_OG0000055_p32_OG0000061_p33_OG0000062_p48_OG0000077_p49_OG0000078,
    LG+G4: p4_OG0000033_p10_OG0000039_p12_OG0000041_p34_OG0000063,
    LG+F+I+G4: p5_OG0000034_p15_OG0000044_p17_OG0000046_p18_OG0000047_p21_OG0000050_p23_OG0000052_p25_OG0000054_p27_OG0000056_p35_OG0000064_p36_OG0000065_p38_OG0000067_p39_OG0000068_p43_OG0000072,
    LG+G4: p6_OG0000035_p30_OG0000059,
    LG+G4: p8_OG0000037_p19_OG0000048_p42_OG0000071_p47_OG0000076,
    LG+G4: p14_OG0000043_p31_OG0000060_p46_OG0000075,
    LG+I+G4: p20_OG0000049_p45_OG0000074,
    LG+F+G4: p24_OG0000053,
    LG+I+G4: p28_OG0000057_p29_OG0000058,
    LG+G4: p37_OG0000066_p40_OG0000069_p44_OG0000073;
end;

varMerged=$(grep "Bayesian information criterion (BIC) score:" Analyses/IQ-TREE/spp_MergedPartitions.iqtree | cut -d" " -f6
echo -e "$varSP\n$varSPP\n$varMerged"
481216.3913
478647.5903
477711.6257
```

At this point you should have your best-fit partitioning scheme and corresponding evolutionary models. We are now ready for the next step, the tree inference...

>Precomputed results can be found [here](https://github.com/for-giobbe/phy/blob/master/2021/Analyses/Precomputed_IQ-TREE.tar.gz).

---

### NOTES:

Traditional model relyes on the assumption that evolution of DNA sequences follow a markov process. However, again, it may not **always** be true. For this reason IQ-TREE implements three *matched-pairs tests of symmetry* ([Naser-Khdour et al., 2019](https://academic.oup.com/gbe/article/11/12/3341/5571717)) to test the two assumptions of *stationarity* and *homogeneity*. Morever are also implemented some Lie Markov models ([Woodhams et al., 2015](https://academic.oup.com/sysbio/article/64/4/638/1650486)), many of which are non-reversible models.

Finally, in Bayesian Inference in the latest years has become very fashionable the CAT model. This mixture model allows **each position** on the alignment to evolve differently from the others, whithout needing of any *a priori* knowledge (*i.e* partitioning scheme). IQ-TREE implements a ML variant called **GHOST**([Crotty et al., 2020](https://academic.oup.com/sysbio/article-abstract/69/2/249/5541793?redirectedFrom=fulltext)). However they are very parameters rich and can only be used when analyzing a lot of data (*i.e* phylogenomic).

## further reading: 

Some great [basic](http://www.iqtree.org/doc/Tutorial) and [advanced](http://www.iqtree.org/doc/Advanced-Tutorial) tutorials are available from the authors themselves on ModelFinder and IQ-TREE.

[Here](http://www.iqtree.org/doc/Complex-Models) is a very nice tutorial on complex models in IQ-TREE.

[Here](http://www.iqtree.org/doc/Frequently-Asked-Questions) you'll find few words on the composition test carried out by IQ-TREE.

[Interesting paper on how concatenation/coalescence impacts mammalian phylogeny](https://onlinelibrary.wiley.com/doi/full/10.1111/cla.12170?casa_token=X0ctrSm4S1AAAAAA%3AgiB9v0MtJDO6vMWOigdvW9JrgYuJTebMen6zYxg9S0nP8MWIi2zA2fwWfi-lJlMCD9Ir1MDCzkBeyVwg).

[Interesting paper on how systematic errors as heterotachy impact plants phylogeny](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3237385/pdf/evr105.pdf)

